File: gui/hello/hello.py
#! /usr/bin/python

# initialize toolkit
import pygtk
pygtk.require('2.0')
import gtk

class Hello(gtk.Window):
    def __init__(self, title):
        ''' build GUI '''
        gtk.Window.__init__(self)
        # other attributes, e.g. title, border, default-size, etc
        self.set_title('%s: %s' % (self.__class__.__name__, title))

        # build GUI: create, connect-signal, pack
        self.button = gtk.Button('Hello world')
        self.button.connect('clicked', self.on_hello_clicked)
        self.add(self.button)

        # other signals
        self.connect('destroy', self.on_destroy)
        self.connect('delete-event', self.on_delete_event)

        # show GUI
        self.set_size_request(200, 100)
        self.show_all()

    def run(self):
        ''' toolkit framework event-loop '''
        gtk.main()

    def on_destroy(self, widget, data=None):
        ''' sample of signal callback '''
        print('on_destroy')
        gtk.main_quit()

    def on_delete_event(self, widget, event, data=None):
        ''' sample of event callback '''
        print('on_delete_event')
        return False    # keep event bubbling

    def on_hello_clicked(self, widget, data=None):
        ''' sample of control signal callback '''
        print('on_hello_clicked')

# start of application of unit test
if __name__ == '__main__':
    app = Hello('Demo')
    app.run()

File: gui/hello/hello.c
#include <gtk/gtk.h>

static void cb_destroy(GtkWidget*, gpointer);
static gboolean cb_delete_event(GtkWidget*, GdkEvent*, gpointer);

// this is just an example
// client data passed to signal/event callback can be anything
typedef struct {
    char* msg;
} ClientData;

int main(int argc, char* argv[])
{
    GtkWidget *window;
    GtkWidget *label;
    ClientData data = { "data-passed-to-signal" };

    gtk_init(&argc, &argv);     // initialize gtk toolkit

    // build GUI
    window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title(GTK_WINDOW(window), "Hello World!");
    gtk_container_set_border_width(GTK_CONTAINER(window), 10);
    gtk_widget_set_size_request(window, 200, 100);

    label = gtk_label_new("Hello world");
    gtk_container_add(GTK_CONTAINER(window), label);

    // connect signals and event (if applicable)
    g_signal_connect(G_OBJECT(window), "destroy",
                G_CALLBACK(cb_destroy), &data);
    g_signal_connect(G_OBJECT(window), "delete-event",
                G_CALLBACK(cb_delete_event), NULL);

    // show widgets tree and fall into event loop
    gtk_widget_show_all(window);
    gtk_main();

    return 0;
}

static void
cb_destroy(GtkWidget *widget, gpointer client_data)
{
    ClientData *pdata = (ClientData*) client_data;
    g_print("destroy signal: %s\n", pdata ? pdata->msg : "");
    gtk_main_quit();    // break the current event loop
}

static gboolean
cb_delete_event(GtkWidget *widget, GdkEvent *event, gpointer client_data)
{
    g_print("Delete-event\n");
    sleep(1);           // let you see clearly the callback sequence

    return FALSE;       // continue event-process-chain
    // default delete-event-process on window is to emit "destroy" signal
}

